handler.go

```diff
@@ -0,0 +1,276 @@
+// Copyright 2015 Canonical Ltd.
+// Licensed under the LGPLv3, see LICENCE file for details.
+
+package httprequest
+
+import (
+	"encoding/json"
+	"net/http"
+	"reflect"
+
+	"github.com/julienschmidt/httprouter"
+	"gopkg.in/errgo.v1"
+)
+
+// ErrorMapper holds a function that can convert a Go error
+// into a form that can be returned as a JSON body from an HTTP request.
+// The httpStatus value reports the desired HTTP status.
+type ErrorMapper func(err error) (httpStatus int, errorBody interface{})
+
+var (
+	paramsType             = reflect.TypeOf(Params{})
+	errorType              = reflect.TypeOf((*error)(nil)).Elem()
+	httpResponseWriterType = reflect.TypeOf((*http.ResponseWriter)(nil)).Elem()
+	httpHeaderType         = reflect.TypeOf(http.Header(nil))
+)
+
+// Handle converts a function into an httprouter.Handle. The argument f
+// must be a function of one of the following three forms, where ArgT
+// must be a struct type acceptable to Unmarshal and ResultT is a type
+// that can be marshaled as JSON:
+//
+//	func(w http.ResponseWriter, p Params, arg *ArgT)
+//	func(w http.ResponseWriter, p Params, arg *ArgT) error
+//	func(header http.Header, p Params, arg *ArgT) (ResultT, error)
+//
+// When processing a call to the returned handler, the provided
+// parameters are unmarshaled into a new ArgT value using Unmarshal,
+// then f is called with this value. If the unmarshaling fails, f will
+// not be called and the unmarshal error will be written as a JSON
+// response.
+//
+// If an error is returned from f, it is passed through the error mapper before
+// writing as a JSON response.
+//
+// In the third form, when no error is returned, the result is written
+// as a JSON response with status http.StatusOK.
+//
+// Handle will panic if the provided function is not in one
+// of the above forms.
+func (e ErrorMapper) Handle(f interface{}) httprouter.Handle {
+	fv := reflect.ValueOf(f)
+	pt, err := checkHandleType(fv.Type())
+	if err != nil {
+		panic(errgo.Notef(err, "bad handler function"))
+	}
+	return e.handleResult(fv.Type(), handleParams(fv, pt))
```

Comment:
To phrase things another way, the only error we can get here is if the function type isn't
correct. All other errors happen when the HTTP request is being services (for example
when parameters cannot be unmarshaled)